### DISCUSSION TITLE
DEVICE_INIT_DT_GET macro expands to undeclared

### DISCUSSION THREAD

**Discussion by HenrikKaa** [zephyrproject-rtos/zephyr#66662] [Q&A]
**Describe the bug**
I'm trying to get the `init_entry` of a device using `DEVICE_INIT_DT_GET` macro but when compiling the resulting `__init___device_dts_ord_N` is undeclared.

In my specific case
`DEVICE_DT_GET(DT_NODELABEL(spiflash))` expands to `&__device_dts_ord_104` (which is correct)
`DEVICE_INIT_DT_GET(DT_NODELABEL(spiflash))` expands to `&__init___device_dts_ord_104` (which is undeclared)

The grep result from build folder (without the offending lines)

```
grep -r "__init___device_dts_ord_104"
grep: build/zephyr/zephyr_pre0.elf: binary file matches
grep: build/zephyr/zephyr.elf: binary file matches
grep: build/zephyr/drivers/flash/CMakeFiles/drivers__flash.dir/spi_nor.c.obj: binary file matches
grep: build/zephyr/drivers/flash/libdrivers__flash.a: binary file matches
grep: build/zephyr/zephyr_pre1.elf: binary file matches
```

So clearly the init function does exist but something goes wrong compiling. The actual init function can also be found in zephyr.map if I remove the offending lines.

```
 .z_init_POST_KERNEL80_
                0x0000000010022cf8        0x8 zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj)
```

**To Reproduce**
I can't say more than try to use the macro, it does not work for any device in my project. However, `tests/lib/devicetree/devices` runs just fine.

**Expected behavior**
Get reference to the devices init_entry.

**Impact**
Major annoyance

**Logs and console output**
```
/.../zephyr/include/zephyr/init.h:107:44: error: '__init___device_dts_ord_104' undeclared (first use in this function); did you mean '__device_dts_ord_104'?
  107 | #define Z_INIT_ENTRY_NAME(init_id) _CONCAT(__init_, init_id)
```

**Environment (please complete the following information):**
 - OS: Linux (Ubuntu 22.04)
 - Zephyr SDK 0.16.1
 - Zephyr 3.3.99

**Comment by gmarull** [comment]
Provided tests/lib/devicetree/devices runs just fine, it looks like a problem on your setup. I'd also discourage usage of this API, not even sure why it's there.

**Reply by HenrikKaa** [reply]
I think it is the only way to run initialisation functions manually until this progresses https://github.com/zephyrproject-rtos/zephyr/issues/39896

**Comment by aasinclair** [comment]
This looks to be caused by _static_ here:
https://github.com/zephyrproject-rtos/zephyr/blob/38010c111e2342e0c574ee88c8704a39ee7e6006/include/zephyr/device.h#L972

The unit test works because the device is declared in the same file.
